NewForce-Cohort-9/tabloid-full-stack-full-stack-force
# Finished Ticket #45 - Double Admin

# Description
Added two-admin protection to some actions. This adds another table to the DB in order to complete ticket requirements. It makes it so that two admins have to agree to deactivate another admin account or change an admin's user type. 

[Trello Ticket #45][(https://trello.com/c/DqjMk57X](https://trello.com/c/BIIvafGW))

# Testing Instructions
Add and commit any unsaved changes on your branch. Run git checkout main, git fetch --all, and `git checkout two-admins-del`.
If you don't see the most recent version of my files, run `git pull origin two-admins-del`.

First, you will need to create a new table with the 05 SQL script included with this commit. Or, copy the code here: 

```
CREATE TABLE AdminActionRequest (
[Id] integer PRIMARY KEY IDENTITY,
[TargetUserId] integer NOT NULL, 
[RequestingAdminId] integer NOT NULL,
[ApprovingAdminId] integer, 
[ActionType] nvarchar(255) NOT NULL,
[IsCompleted] bit NOT NULL DEFAULT 0, 
[RequestedAt] datetime NOT NULL, 
[ApprovedAt] datetime

CONSTRAINT FK_AdminActionRequest_TargetUser FOREIGN KEY (TargetUserId)
REFERENCES UserProfile(Id),

CONSTRAINT FK_AdminActionRequest_RequestingAdmin FOREIGN KEY (RequestingAdminId)
REFERENCES UserProfile(Id),

CONSTRAINT FK_AdminActionRequest_ApprovingAdmin FOREIGN KEY (ApprovingAdminId)
REFERENCES UserProfile(Id),
);

--The below enforces uniqueness where a request IS NOT completed.
--In order to avoid confusion, for a single action type (Deactivating, ChangeUserType, etc),
--only 1 non-completed admin request can be opened for a single target user with the same action type.
--so for targetUserId 1, you could only have 1 "Deactivating" entry if the entry IS NOT completed.

CREATE UNIQUE INDEX IX_TargetUser_ActionType 
ON AdminActionRequest (TargetUserId, ActionType)
WHERE IsCompleted = 0;
```

See the comments in the code above for an explanation. Once you have the code copied, go to Visual Studio (purple), find your Tabloid database, and right click it to create a new query. Within the new query, paste the code that you copied above and click the green the button to execute the query. Your database should have updated with a new table. You can run a `SELECT * FROM AdminActionRequest` query to ensure that the table was created.

Afterwards, inside of Visual Studio (purple), click the execute with debugger solid green button to launch the app.
In your terminal, navigate to where your package.json file is in you client folder. Run "npm start" in your terminal.


In order to test this, you need to have 3-4 admin accounts, so if you don't, go to User Profiles list and change the user type of the accounts until you have 3-4 accounts that are admins. Once you have that, follow the instructions below:

1. Make sure that you are logged in as an admin account. One admin I have been using is `foo@bar.com`. Another admin I have been using is `gteanby6@craigslist.orgx`. Pick out 2 admin accounts, one that you are logged in as already, and another admin account.
2. Navigate to user profiles list by clicking User Profiles in navbar, `/profiles` route. 
3. In the user profile list, find another Admin account (different from the 2 you picked out in step 1).  Find an admin account that has a "Deactivate" button next to it. If you don't see one with a Deactivate button, click on the Reactivate button for an admin account and it'll make it so that you can now Deactivate.
4. Click on Deactivate button for the admin account that you have chosen to attempt to deactivate. When you are taken to the confirmation route, `/profile/deactivate/{id}`, click the solid red "Deactivate" button. 
5. You should now see a window alert message after clicked that says "Request sent and will wait for another admin to approve". No other further action should have happened. 
6. Back on User Profile List, remember the account that you attempted to deactivate. You will now log out and log back in 
as the other admin account that you picked out. In my case, I switched from foo@bar.com to gteanby6@craiglist.orgx.
7. While logged in as a second admin, go to User Profile List, and on the same user as before, go through the Deactivate process. This time when you click the solid red Deactivate button, you should see a window.alert message that says "Approved admin action" since you now have 2 admin approvals for this deactivation. The account should have deactivated. 

Once you have done that, let's make sure that changing the user type of an admin is also gated with this approval process.

While still logged in as an admin, from the user profile list, find another account that has a UserType of Admin that is different from the 2 admins that you are logging in as. This time, click on the "Edit" button next to this user's name.  Attempt to change this user's type from "Admin" to "Author" and click Save. You should be given the same window.alert message from earlier, that the request is sent and will wait for another admin to approve.

Remembering the user that you are editing, log out and log back in as your 2nd admin account. Go back to User Profile List and find the user that you are editing. This time, as the 2nd admin, when you go to change the user type from Admin to Author, you should see a window alert "Approved admin action" and it should change the user's type to Author.

Other things:
From User Profile list, pick a user that has a user type of Author. Click Edit, and change their user type to Admin. Ensure that this process still works without the approval system, and that nothing else is effected. Ensure users can be reactivated as normal, etc. 

Double admin request process should only apply in these scenarios:
1) Changing an Admin's user type from Admin to author
2) Deactivating an Admin account

Otherwise, user profile list should behave as before without the double admin.


# Checklist:
[x] My code follows the style guidelines of this project
[x] I have performed a self-review of my code
[x] I have commented my code, particularly in hard-to-understand areas
[x] My changes generate no new warnings or errors
[x] I have added test instructions that prove my fix is effective or that my feature works
